src/sqlfluff/core/rules/base.py

```diff
@@ -484,149 +586,107 @@ def _eval(self, context: RuleContext) -> EvalResultType:
+            new_fixes: List[LintFix] = []
+
+            if res is None:
+                # Assume this means no problems (also means no memory)
+                pass
+            elif isinstance(res, LintResult):
+                # Extract any memory
+                memory = res.memory
+                self._adjust_anchors_for_fixes(context, res)
                 self._process_lint_result(
-                    elem, templated_file, ignore_mask, new_lerrs, new_fixes
+                    res, templated_file, ignore_mask, new_lerrs, new_fixes
                 )
-        else:  # pragma: no cover
-            raise TypeError(
-                "Got unexpected result [{!r}] back from linting rule: {!r}".format(
-                    res, self.code
+            elif isinstance(res, list) and all(
+                isinstance(elem, LintResult) for elem in res
+            ):
+                # Extract any memory from the *last* one, assuming
+                # it was the last to be added
+                memory = res[-1].memory
+                for elem in res:
+                    self._adjust_anchors_for_fixes(context, elem)
+                    self._process_lint_result(
+                        elem, templated_file, ignore_mask, new_lerrs, new_fixes
+                    )
+            else:  # pragma: no cover
+                raise TypeError(
+                    "Got unexpected result [{!r}] back from linting rule: {!r}".format(
+                        res, self.code
+                    )
                 )
-            )
 
-        for lerr in new_lerrs:
-            self.logger.debug("!! Violation Found: %r", lerr.description)
-        for fix in new_fixes:
-            self.logger.debug("!! Fix Proposed: %r", fix)
-
-        # Consume the new results
-        vs += new_lerrs
-        fixes += new_fixes
-
-        # The raw stack only keeps track of the previous raw segments
-        if len(segment.segments) == 0:
-            raw_stack += (segment,)
-        # Parent stack keeps track of all the parent segments
-        parent_stack += (segment,)
-
-        for idx, child in enumerate(segment.segments):
-            dvs, raw_stack, child_fixes, memory = self.crawl(
-                segment=child,
-                ignore_mask=ignore_mask,
-                parent_stack=parent_stack,
-                siblings_pre=segment.segments[:idx],
-                siblings_post=segment.segments[idx + 1 :],
-                raw_stack=raw_stack,
-                memory=memory,
-                dialect=dialect,
-                fname=fname,
-                fix=fix,
-                templated_file=templated_file,
-            )
-            vs += dvs
-            fixes += child_fixes
-        return vs, raw_stack, fixes, memory
+            for lerr in new_lerrs:
+                self.logger.debug("!! Violation Found: %r", lerr.description)
+            for lfix in new_fixes:
+                self.logger.debug("!! Fix Proposed: %r", lfix)
```

Comment:
Renamed `fix` to `lfix` to avoid type error with the differently typed `fix` parameter.